src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -108,4 +136,281 @@ protected int getCacheLimit() {
+        protected DynamicObject executeMatch(DynamicObject regexp, DynamicObject string, Matcher matcher,
+                int startPos, int range, boolean onlyMatchAtStart,
+                @Cached("createBinaryProfile()") ConditionProfile matchesProfile) {
+            assert RubyGuards.isRubyRegexp(regexp);
+            assert RubyGuards.isRubyString(string);
+
+            if (getContext().getOptions().REGEXP_INSTRUMENT_MATCH) {
+                instrument(regexp, string, onlyMatchAtStart);
+            }
+
+            int match = runMatch(matcher, startPos, range, onlyMatchAtStart);
+
+            if (matchesProfile.profile(match == -1)) {
+                return nil();
+            }
+
+            assert match >= 0;
+
+            final Region region = matcher.getEagerRegion();
+            final DynamicObject dupedString = (DynamicObject) dupNode.call(null, string, "dup");
+            DynamicObject result = allocateNode.allocate(matchDataClass(), Layouts.MATCH_DATA.build(dupedString,
+                    regexp, region, null));
+            return (DynamicObject) taintResultNode.maybeTaint(string, result);
+        }
+
+        @TruffleBoundary
+        private int runMatch(Matcher matcher, int startPos, int range, boolean onlyMatchAtStart) {
+            // Keep status as RUN because MRI has an uninterruptible Regexp engine
+            if (onlyMatchAtStart) {
+                return getContext().getThreadManager().runUntilResultKeepStatus(this,
+                        () -> matcher.matchInterruptible(startPos, range, Option.DEFAULT));
+            } else {
+                return getContext().getThreadManager().runUntilResultKeepStatus(this,
+                        () -> matcher.searchInterruptible(startPos, range, Option.DEFAULT));
+            }
+        }
+
+        @TruffleBoundary
+        protected void instrument(DynamicObject regexp, DynamicObject string, boolean fromStart) {
+            Rope source = Layouts.REGEXP.getSource(regexp);
+            Encoding enc = Layouts.STRING.getRope(string).getEncoding();
+            RegexpOptions options = Layouts.REGEXP.getOptions(regexp);
+            MatchInfo matchInfo = new MatchInfo(new RegexpInfo(source, enc, options), fromStart);
+            ConcurrentOperations.getOrCompute(matchedRegexps, matchInfo, x -> new AtomicInteger()).incrementAndGet();
+        }
+
+        protected DynamicObject matchDataClass() {
+            return getContext().getCoreLibrary().getMatchDataClass();
+        }
+    }
+
+    private static class MatchInfo {
+        private final RegexpInfo regexpInfo;
+        private final boolean matchStart;
+
+        MatchInfo(RegexpInfo regexpInfo, boolean matchStart) {
+            this.regexpInfo = regexpInfo;
+            this.matchStart = matchStart;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + (matchStart ? 1231 : 1237);
```

Comment:
Where do these numbers come from?

Reply:
The methods were auto generated by eclipse.

Reply:
Oh, I never used that.
Not sure how 1231/1237 are better than 1 or 2, though.
I think it's worth cleaning up a bit in `equals()` at least.

Reply:
They appear to have been picked to match`Boolean.hashCode()` in the jdk. They were probably picked as simple two sufficiently large primes so that they won't cause two many collisions when combined into larger hashes.

Reply:
I see, I think it would be cleaner to use `Boolean.hashCode(matchStart)` instead of the ternary if with hard-coded values.